service/Dockerfile

```diff
@@ -0,0 +1,78 @@
+# LS docker file
+
+# build via (docker root = project root):
+# docker build -t absaoss/login-service:latest \
+# --build-arg BUILD_PROXY=http://my.cool.proxy.here:3128 \
+# --build-arg CONFIG=./path/to/my.awesome.local.application.yaml \
+# --build-arg SSL=true \
+# --build-arg SSL_DNAME="CN=*.my.domain.com, OU=project1, O=mycorp, L=Johannesburg, ST=Gauteng, C=za" .
+# run via
+# docker run -p 8080:8080 -p 8443:8443 absaoss/login-service:latest
+
+# test via:
+# http://localhost:8080/token/public-key
+# https://localhost:8443/token/public-key
+
+
+# --- Base image part (common for SSL true|false) ---
+# Conditional Docker image creation idea: https://stackoverflow.com/a/60820156/1773349
+# change to OFF to disable SSL
+ARG SSL=true
+
+# Tomcat OS base image, feel free to use another (e.g. your hardened one)
+ARG BASE_IMAGE=tomcat:9-jdk8-corretto
+FROM $BASE_IMAGE as base
+
+LABEL org.opencontainers.image.authors="ABSA"
+
+ARG BUILD_PROXY
+ARG CONFIG=./src/main/resources/example.application.yaml
+
+ENV LS_WAR_PATH=./target/scala-2.12
+# deploy as root
+ENV LS_WAR=ROOT.war
+
+# Copy Spring application properties
+COPY $CONFIG /opt/application.yaml
+
+COPY $LS_WAR_PATH/*.war /usr/local/tomcat/webapps/$LS_WAR
+
+ENV SPRING_CONFIG_ADDITIONAL-LOCATION=/opt/application.yaml
+
+ENV http_proxy=$BUILD_PROXY
+ENV https_proxy=$BUILD_PROXY
+ENV HTTP_PROXY=$BUILD_PROXY
+ENV HTTPS_PROXY=$BUILD_PROXY
+
+# uncomment and add packages you would need to install via yum if any
+#RUN yum -y update
+#RUN yum -y install htop procps
+
+# --- SSL=true image specifics ---
+FROM base AS base-ssl-true
+ENV SSL_ENABLED=true
+RUN echo "This stage sets SSL=$SSL_ENABLED"
+# Copy Tomcat properties (to enable SSL)
+COPY ./src/main/resources/tomcat-ssl/server.xml /usr/local/tomcat/conf/
+
+ARG SSL_DNAME="CN=*.my.example.com, OU=project1, O=yourcompany, L=Johannesburg, ST=Gauteng, C=za"
+
+# A self-seigned certificate for HTTPS
+RUN keytool -genkeypair -keyalg RSA -alias tomcat -keysize 2048 \
+    -dname "$SSL_DNAME" \
+    -validity 365 -storepass changeit -keystore /usr/local/tomcat/conf/selfsigned.p12 -storetype PKCS12
+
+EXPOSE 8080 8443
+CMD ["catalina.sh", "run"]
+
+# --- SSL=false image specifics ---
+FROM base AS base-ssl-false
+ENV SSL_ENABLED=false
+RUN echo "This stage sets SSL=$SSL_ENABLED"
+
+EXPOSE 8080
+CMD ["catalina.sh", "run"]
+
+# --- Final image assembly ---
+FROM base-ssl-${SSL} AS final
+RUN echo "Running final stage with SSL_ENABLED=$SSL_ENABLED."
```

Comment:
since this is a multi-stage docker build, don't you need to copy some app/config files into the last build(s) ? I.e. after the last FROM (actually two of them obviously can be 'final')

E.g. these files 
 
```
# Copy Spring application properties
COPY $CONFIG /opt/application.yaml

COPY $LS_WAR_PATH/*.war /usr/local/tomcat/webapps/$LS_WAR
```

I haven't tried it though, just did the review. If you want me to build it on my Mac, please tell me

Reply:
(If I understood and remember multi-stage docker builds correctly, every FROM is a new layer and doesn't contain artifacts from the previous layers - ENV vars and files?)

Reply:
I think your comment would be correct in the case of multi-stage docker build that do not use each other as bases (as explained at e.g. https://www.baeldung.com/linux/docker-copy-variables-stages) - there, the extra copying and redefinitions of ARGs would be needed (while ENV being used throughout the whole file.

However, here, the `base` image is used via `FROM` in subsequent images and so on, creating a chain, so no extra copying or ARG redefinition should be needed. If it was needed and missing, the solution would not work properly and would the `final` image would be oblivious to ARGs specified for the `base` image -- but it works correctly, thus serving as evidence of no need of copy/ARG-redef in this case. It is somehow outlined at https://stackoverflow.com/questions/57577265/which-dockerfile-instructions-are-inherited-in-deriving-images where such an _image inheritance_ is discussed.

Reply:
Okay, I didn't now about these particular inheritance capabilities. Cool, thanks :)